### PR TITLE
feat: use Lake.withLockFile in cache

### PR DESCRIPTION
**Pull request by kim-em**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
It might be a while before we can merge this!

**Comment by tydeu** [comment]
@semorrison Why does this need a Lean dump? `lake.lock` landed before the current `08-15` nightly used mathlib?

**Comment by semorrison** [comment]
> @semorrison Why does this need a Lean bump? `lake.lock` landed before the current `08-15` nightly used by mathlib?

Absolutely right, I must not have branched off the latest master on my first attempt at this. I will force push this now.

**Comment by alreadydone** [comment]
My PR [fails CI run](https://github.com/leanprover-community/mathlib4/actions/runs/5887436948/job/15966827855?pr=6625) and the only error (I think) concerns `lake.lock`. Do you know what's happening? Thanks.

```
[3800/3804] Building Mathlib.NumberTheory.ZetaFunction
[3802/3804] Building Mathlib.AlgebraicGeometry.Morphisms.FiniteType
[3803/3804] Building Mathlib
error: no such file or directory (error code: 2)
  file: ./build/lake.lock
Error: The process '/usr/bin/bash' failed with exit code 1
Remove matcher: gcc
```

**Comment by tydeu** [comment]
@semorrison One issue that just occurred to me with this PR is that `cache` lock will conflict with Lake's in #6603. Thus, there needs to be a `--no-lock` flag in cache (or something similar) to disable locking when called from within the build process.

**Comment by semorrison** [comment]
I've implemented `--no-lock`.

Really we should be using the `lean4-cli` library here, but as we want to replace `cache` anyway I don't think it's worth doing, so I've just added to step to the argument mangling.

File: Cache/Main.lean [hunk 1]
```diff
@@ -61,12 +61,21 @@ def curlArgs : List String :=
...
   if leanTarArgs.contains (args.headD "") then validateLeanTar
+  -- An optional argument `--no-lock` disables use of Lake's lock file mechanism.
+  -- This is required if `lake` is calling `cache` itself.
+  let (args, noLock) := if "--no-lock" ∈ args then
+    (args.erase "--no-lock", true)
+  else
+    (args, false)
+  -- For now we hard code the build directory, to avoid having to process the lakefile.
+  -- Perhaps if https://github.com/leanprover/lake/issues/176 is addressed this can be revisited.
+  (if noLock then id else Lake.withLockFile (FilePath.mk "build" / "lake.lock")) do
```
Comment (tydeu) [inline Cache/Main.lean:78]:
```suggestion
  (if noLock then id else Lake.withLockFile (FilePath.mk "build" / Lake.lockFileName)) do
```

Might be worth it.

**Review by tydeu — APPROVED** [review]
LGTM.

**Comment by semorrison** [comment]
Marking as WIP, as we have temporarily reverted the lockfile.

**Pull request closed without merging**